typing/typecore.ml

```diff
@@ -845,6 +862,40 @@ let wrap_disambiguate msg ty f x =
   | Wrong_name_disambiguation (env, wrong_name) ->
     raise (Error (wrong_name.name.loc, env, Wrong_name (msg, ty, wrong_name)))
 
+(* Use a dummy instance of [NameChoice] to disambiguate between [float array]
+   and [floatarray]. Note that an actual ambiguous choice is not possible, as we
+   default to ['a array] in case of doubt. *)
+
+module FloatArray = NameChoice(struct
+    type t = bool
+    type usage = unit
+    let kind = Datatype_kind.Record
+    let get_name = function
+      | false -> "array"
+      | true -> "floatarray"
+    let get_type = function
+      | false -> Predef.type_array Predef.type_unit
+      | true -> Predef.type_floatarray
+    let lookup_all_from_type _loc () p _env =
+      [Path.same p Predef.path_floatarray, ignore]
+    let in_env _ = true
+  end)
+
+let disambiguate_array_type loc env expected_ty =
+  let lid = Location.mkloc (Longident.Lident "#array#") loc in
+  let expected_type, candidates =
```

Comment:
The candidates should always be the equivalent of `Ok [false, ignore; true, ignore]`, they should not depend on whether there is an expected type.